### ISSUE TITLE
Why  all_fea = torch.cat((all_fea, torch.ones(all_fea.size(0), 1)), 1)  ?

### ISSUE THREAD

**Issue by YYShi12** [tim-learn/SHOT#3]
Hi, I can't understand why the feature embedding is added one dimension whose value is 1 when obtaining the pseudo label.

Would you please explain it? 

Thanks

**Comment by tim-learn** [comment]
> Hi, I can't understand why the feature embedding is added one dimension whose value is 1 when obtaining the pseudo label.
> Would you please explain it?
> Thanks

This operation is done to be consistent with the last layer with bias=True.

**Comment by YYShi12** [comment]
Thanks. I've understood this operation.

But I can't understand another operation, all_fea = (all_fea.t() / torch.norm(all_fea, p=2, dim=1)).t() . 
Why feature embeddings are divided by their L2 norm?

**Comment by tim-learn** [comment]
> Thanks. I've understood this operation.
> But I can't understand another operation, all_fea = (all_fea.t() / torch.norm(all_fea, p=2, dim=1)).t() .
> Why feature embeddings are divided by their L2 norm?

We use the cosine distance to measure the similarity. For simplicity, we first l2-normalize the feature vector.

**Comment by YYShi12** [comment]
Thanks a lot for your prompt reply.
